firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,301 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
-                                } else {
-                                    // branch with NO children, not at the top
-                                    Poll::Ready(Some(Err(api::Error::InternalError(Box::new(
-                                        MerkleError::ParentLeafBranch,
-                                    )))))
-                                };
-                            };
-
-                            let next = self
-                                .merkle
-                                .get_node(leftmost_address)
-                                .map_err(|e| api::Error::InternalError(Box::new(e)))?;
-
-                            parents.push((last_node, leftmost_position as u8));
-
-                            last_node = next;
-                        }
-                        NodeType::Leaf(leaf) => break leaf,
-                        NodeType::Extension(_) => todo!(),
-                    }
-                };
 
-                // last_node should have a leaf; compute the key and value
-                let current_key = key_from_parents_and_leaf(&parents, leaf);
+                // always put the sentinal node in parents
+                let mut parents = vec![(root, 0)];
+
+                let next_result = find_next_result(merkle, &mut parents)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))
+                    .transpose();
 
-                self.key_state = IteratorState::Iterating { last_node, parents };
+                self.key_state = IteratorState::Iterating { parents };
 
-                current_key
+                Poll::Ready(next_result)
             }
+
             IteratorState::StartAtKey(key) => {
-                // TODO: support finding the next key after K
-                let root_node = self
-                    .merkle
-                    .get_node(self.merkle_root)
+                if key.is_empty() {
+                    self.key_state = IteratorState::StartAtBeginning;
+                    return self.poll_next(_cx);
+                }
+
+                let root_node = merkle
+                    .get_node(*merkle_root)
                     .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                let (found_node, parents) = self
-                    .merkle
+                let (found_node, mut parents) = merkle
                     .get_node_and_parents_by_key(root_node, &key)
                     .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                let Some(last_node) = found_node else {
-                    return Poll::Ready(None);
-                };
+                if let Some(found_node) = found_node {
+                    let value = match found_node.inner() {
+                        NodeType::Branch(branch) => branch.value.as_ref(),
+                        NodeType::Leaf(leaf) => Some(&leaf.data),
+                        NodeType::Extension(_) => None,
```

Comment:
Q: Is it possible that `found_node` is actually a `NodeType::Extension`? If so, under what conditions? Can we add a test case for that?

Reply:
The simple answer is no, `found_node` can't be an extension. But it is problematic when the key lands us on an extension and I have added a test case for it.